### PR TITLE
feature: Timestamp queries

### PR DESCRIPTION
**Pull request by FL33TW00D**
**Checklist**

- [X] Run `cargo clippy`.
- [X] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
https://github.com/gfx-rs/wgpu/pull/2528/files
https://github.com/gpuweb/gpuweb/issues/1325
https://github.com/gpuweb/gpuweb/issues/2046
https://github.com/gfx-rs/metal-rs/pull/263

**Description**
This is a WIP implementation of timestamp queries for compute/render passes. Only implemented for Metal and compute for now.

**TODO**
- [x] Release required changes on `metal-rs` [here](https://github.com/gfx-rs/metal-rs/pull/265).
- [X] Implement and test on Metal for render passes.
- [ ] Implement and test for all other back ends for compute and render passes.
- [ ] Resolve issue with attaching the same sample buffer multiple times.
- [X] Update all existing examples using `ComputePassDescriptor` and `RenderPassDescriptor`.

**Testing**
Modified `hello-compute` and `hello-triangle` and successfully recorded the execution time. Only tested on M1 so far.

### CODE REVIEW

File: wgpu-hal/src/metal/command.rs
Comment (FL33TW00D) [inline wgpu-hal/src/metal/command.rs:953]: Any ideas on a good solution for this? We could map from query_set -> sample buffer attachment index?

File: Cargo.toml
Comment (cwfitzgerald) [inline Cargo.toml]: Erroneous patch

File: wgpu-hal/src/metal/command.rs
Comment (cwfitzgerald) [reply wgpu-hal/src/metal/command.rs]: I don't quite get the problem?

**Comment by cwfitzgerald** [comment]
For the other backends, can't we just implement in terms of other existing operations on the command encoder?

File: wgpu/src/lib.rs
Comment (Wumpf) [inline wgpu/src/lib.rs]:
```suggestion
/// https://gpuweb.github.io/gpuweb/#dictdef-gpucomputepasstimestampwrites).
```

File: wgpu/src/lib.rs
Comment (Wumpf) [inline wgpu/src/lib.rs]: uhoh, looks like the spec has moved on and now this is handled without an enum https://gpuweb.github.io/gpuweb/#dictdef-gpucomputepasstimestampwrites
Looks like we now need two `Option<u32>` fields on `ComputePassTimestampWrite`/`RenderPassTimestampWrite`

File: wgpu/src/lib.rs
Comment (Wumpf) [inline wgpu/src/lib.rs]:
```suggestion
/// https://gpuweb.github.io/gpuweb/#dictdef-gpurenderpasstimestampwrites).
```

File: wgpu/src/lib.rs
Comment (Wumpf) [inline wgpu/src/lib.rs]: as below, this is outdated already again :O

File: wgpu/examples/hello-triangle/main.rs
Comment (Wumpf) [inline wgpu/examples/hello-triangle/main.rs]: hello-triangle is tempting as a testing ground but I think the sample should stay clean of that. It's prime function is to show a minimal triangle to new users, so this needs to move somewhere else

File: wgpu/examples/hello-compute/main.rs
Comment (Wumpf) [inline wgpu/examples/hello-compute/main.rs]: same comment as on hello-triangle applies: Timestamps deserve their own example and don't high-jack these minimal-function examples :)

File: wgpu/Cargo.toml
Comment (Wumpf) [inline wgpu/Cargo.toml]: as commented elsewhere, this type is dead by now again

File: wgpu-core/src/command/render.rs
Comment (Wumpf) [inline wgpu-core/src/command/render.rs]: need to add validation as described in https://www.w3.org/TR/webgpu/#abstract-opdef-validate-timestampwrites which should be then shared with compute

File: wgpu-core/src/command/compute.rs
Comment (Wumpf) [inline wgpu-core/src/command/compute.rs]: need to add validation as described in https://www.w3.org/TR/webgpu/#abstract-opdef-validate-timestampwrites which should be then shared with render

**Review by Wumpf — CHANGES_REQUESTED** [review]
super cool PR, looking forward to this. Maybe tomorrow I get finally to picking up the flag there if @FL33TW00D doesn't have more in the pipe already?

Haven't gotten to actually looking at the actual Metal code yet :(.
Sadly the spec moved already on a bit, so things are supposed to look a lil bit different now, put comments at the key place so it's easy to find.

otherwise also on the todo list belong
* update comments on TIMESTAMP_QUERY feature
* implement (i.e. forward things in) web.rs version

**Comment by FL33TW00D** [comment]
@Wumpf Really appreciate you looking through this, I left it as is due to the release of Chrome 113 being a higher priority.

The metal PR is here: https://github.com/gfx-rs/metal-rs/pull/265

There is still a problem with the counter sample buffer not being kept alive long enough.

**Comment by Wumpf** [comment]
@FL33TW00D I copied your branch over to this fork and continued from there - did the move over the (simpler!) WebGPU spec update so far.
If you give me write access to yours I can push there and we can continue here or we create a new PR to commence work there, up to you :)
https://github.com/rerun-io/wgpu/tree/feature/metal-timestamp

**Comment by Wumpf** [comment]
All done!

There is now a separate example demonstrating all kinds of timer queries and I added implementation for Vulkan & DX12. The later I tested on windows (by forcing the respective backend) and Metal got tested on an M1.

Things left to do:
* [x] make CI happy
* [x] get a new Metal version published (@kvark ?)
* [x] reviewing


Note that command encoder timestamps are still not implemented  for Metal. Left some notes on how to proceed there, but this PR got already too big ;)

**Comment by FL33TW00D** [comment]
With the release of `metal-rs` this should be good to go!

File: wgpu-hal/src/metal/device.rs
Comment (teoxoy) [inline wgpu-hal/src/metal/device.rs]: I think we should either upgrade to foreign-types 0.5.0 (https://github.com/gfx-rs/wgpu/pull/3475#issuecomment-1438457541) or remove it from the dependencies and use the re-export everywhere.

Comment (Wumpf) [reply wgpu-hal/src/metal/device.rs]: opted to remove the direct dependency to foreign-types since seems to originate deeper down the tree than Metal. Isn't used for anything else than this here anyways

File: wgpu-hal/src/dx12/command.rs
Comment (cwfitzgerald) [inline wgpu-hal/src/dx12/command.rs]: Why isn't that using write_timestamp like the others?

File: wgpu/examples/timestamp-queries/main.rs
Comment (cwfitzgerald) [inline wgpu/examples/timestamp-queries/main.rs]: This example could likely use some more exposition about what is going on, as someone who has never used them, what information would they need to use them correctly.

File: wgpu/examples/timestamp-queries/main.rs
Comment (cwfitzgerald) [inline wgpu/examples/timestamp-queries/main.rs]: This example should definitely also be tested.

File: wgpu/src/lib.rs
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]:
```suggestion
    /// The index of the query set at which the a start timestamp of this pass is written, if any.
```

File: wgpu/src/lib.rs
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]:
```suggestion
    /// The index of the query set at which the an end timestamp of this pass is written, if any.
```

File: wgpu/src/lib.rs
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]: See above

File: wgpu/src/lib.rs
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]: See above

File: wgpu/src/lib.rs
Comment (cwfitzgerald) [inline wgpu/src/lib.rs]: This shouldn't be in public facing documentation, a comment is fine though.

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]

**Comment by cwfitzgerald** [comment]
Things are looking good, arcinization likely needs to land first though.

**Comment by cwfitzgerald** [comment]
Updated to include new mac CI

**Comment by crowlKats** [comment]
is this close to being done? if so i would start implement the deno part properly

**Comment by Wumpf** [comment]
Fairly close. Api surface won't change: I need to merge main, address the comments and then wait for arcanization to land.
The latter is the reason I didn't hurry too much with it recently.

**Comment by JolifantoBambla** [comment]
Should the web backend be part of this PR?
If so, the `timestamp_writes` members of render and compute passes still need to be mapped and added to their `web_sys` counterparts.
I've updated WebGPU's Web IDL in `web-sys` today, so `GPUComputePassTimestampWrites` and `GPURenderPassTimestampWrites` should be supported in their next release.

**Comment by Wumpf** [comment]
let's put that on a separate PR. this thing is already unwieldy enough as is. But glad to hear we can put that in as well then :)

Not really getting to it very soon sadly. But looks like arcanization needs a bit still anyways..?

**Comment by JolifantoBambla** [comment]
Sounds good :)

**Comment by Wumpf** [comment]
got everything up to date and address review comments but there is some work left to do after adding the tests:
* Mac: @ MoltenVK hard-crashes when it should fail with a known failure. Might be an issue in MoltenVK, need to investigate
* Windows: Turns out that all queries had an undefined behavior issue that is caught by the debug layer when you tried to resolve a queryheap slot to which nothing has been written. WebGPU spec _sounds_ like this should just pass. Which means that we need to track all uninitialized heap spots. Other APIs don't seem to have any issue there, but it's easiest to solve this on the wgpu-core level (and on the flip side adds the kind of overhead to wgpu-hal that we usually don't want to have there)
* Linux: Ci gets stuck on the new test, unclear why

hoping to get to these at some point this week

**Comment by Wumpf** [comment]
the whole "don't copy uninitialized query slots" is a huge rabbit hole. After trying a couple of things I decided do adjourn this, for details see https://github.com/gfx-rs/wgpu/issues/3993

**Comment by Wumpf** [comment]
All green now again!
@cwfitzgerald I know you advocated for putting this in after arcanization, but it's been quite a while and there's a bunch of things that want to be built on top of this PR:
* Solving uninitialized queries
* support inter-pass queries on Metal
* support inter-encoder queries on Metal
* web-backend implementation
* Demo part (@crowlKats )

**Review by Wumpf — APPROVED** [review]

File: examples/timestamp-queries/src/main.rs
Comment (cwfitzgerald) [inline examples/timestamp-queries/src/main.rs]:
```suggestion
    // Instantiates instance of wgpu
```

File: examples/timestamp-queries/src/main.rs
Comment (cwfitzgerald) [inline examples/timestamp-queries/src/main.rs]:
```suggestion
    let mut features = wgpu::Features::TIMESTAMP_QUERY;
```

File: examples/timestamp-queries/src/main.rs
Comment (cwfitzgerald) [inline examples/timestamp-queries/src/main.rs]: Let's cache the features call in a variable, it's not expensive, but not just a getter.

File: wgpu-core/src/command/compute.rs
Comment (cwfitzgerald) [inline wgpu-core/src/command/compute.rs]:
```suggestion
            // Range should always be Some, both values being None should lead to a validation error.
```

**Review by cwfitzgerald — APPROVED** [review]
This is positively amazing work. Short of a couple nits, this is good to land.

**Comment by cwfitzgerald** [comment]
CI is fail due to https://github.com/taiki-e/install-action/issues/178

**Pull request merged**
